iroh-p2p/src/node.rs

```diff
@@ -1095,7 +1209,288 @@ mod tests {
+            }
+            None => {
+                anyhow::bail!("expected NetworkEvent::Gossipsub(Message), received no event");
+            }
+        };
+
+        test_runner_b
+            .client
+            .gossipsub_unsubscribe(topic.clone())
+            .await?;
+        match test_runner_a.network_events.recv().await {
+            Some(NetworkEvent::Gossipsub(GossipsubEvent::Unsubscribed {
+                peer_id,
+                topic: unsubscribe_topic,
+            })) => {
+                assert_eq!(test_runner_b.peer_id, peer_id);
+                assert_eq!(topic, unsubscribe_topic);
+            }
+            Some(n) => {
+                anyhow::bail!("unexpected network event: {:?}", n);
+            }
+            None => {
+                anyhow::bail!("expected NetworkEvent::Gossipsub(Unsubscribed), received no event");
+            }
+        };
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_dht() -> Result<()> {
+        // set up three nodes
+        // two connect to one
+        // try to connect via id
+        let addr_a: Multiaddr = "/ip4/127.0.0.1/tcp/5009".parse().unwrap();
+        let addr_b: Multiaddr = "/ip4/127.0.0.1/tcp/5010".parse().unwrap();
+        let addr_c: Multiaddr = "/ip4/127.0.0.1/tcp/5011".parse().unwrap();
+
+        let cid: Cid = "bafkreieq5jui4j25lacwomsqgjeswwl3y5zcdrresptwgmfylxo2depppq"
+            .parse()
+            .unwrap();
+
+        let addrs = vec![addr_b.clone()];
+
+        let test_runner_a = TestRunnerBuilder::new(addr_a.clone())
+            .no_bootstrap()
+            .build()
+            .await?;
+
+        println!("peer_a: {:?}", test_runner_a.peer_id);
+
+        // peer_id 12D3KooWLo6JTNKXfjkZtKf8ooLQoXVXUEeuu4YDY3CYqK6rxHXt
+        let test_runner_b = TestRunnerBuilder::new(addr_b.clone())
+            .no_bootstrap()
+            .with_seed(ChaCha8Rng::from_seed([0; 32]))
+            .build()
+            .await?;
+
+        println!("peer_b: {:?}", test_runner_b.peer_id);
+
+        let test_runner_c = TestRunnerBuilder::new(addr_c.clone())
+            .no_bootstrap()
+            .with_seed(ChaCha8Rng::from_seed([1; 32]))
+            .build()
+            .await?;
+
+        println!("peer_c: {:?}", test_runner_c.peer_id);
+
+        // connect a and c to b
+        test_runner_a
+            .client
+            .connect(test_runner_b.peer_id, addrs.clone())
+            .await?;
+        test_runner_c
+            .client
+            .connect(test_runner_b.peer_id, addrs.clone())
+            .await?;
+
+        // give time for dht records to update
+        tokio::time::sleep(Duration::from_millis(500)).await;
+
+        // c start providing
+        test_runner_c.client.start_providing(&cid).await?;
+
+        // when `start_providing` waits for the record to make it to the dht
+        // we can remove this `sleep` call
+        tokio::time::sleep(Duration::from_millis(500)).await;
```

Comment:
This could be improved right now by polling repeatedly until either the stuff is found on the DHT or it times out, with a large timeout that certainly won't be too short regardless of how slow the machine is on which the tests run.